java-frontend/src/main/java/org/sonar/java/SonarComponents.java

```diff
@@ -494,67 +493,75 @@ public InputComponent project() {
+    undefinedTypes.stream().forEach(problem -> {
+      List<String> filesAffectedByProblem = problemsToFilePaths.computeIfAbsent(problem, key -> new ArrayList<>());
+      filesAffectedByProblem.add(pathToFile);
+    });
   }
 
   public void logUndefinedTypes() {
-    if (!undefinedTypes.isEmpty()) {
-      javaClasspath.logSuspiciousEmptyLibraries();
-      if (!isAutoScan()) {
-        // In autoscan, test + main code are analyzed in the same batch, and we do not make the distinction between
-        // test and main libraries, everything is inside "sonar.java.libraries", it is expected to let the test property empty.
-        javaTestClasspath.logSuspiciousEmptyLibraries();
-      }
-      logUndefinedTypes(LOGGED_MAX_NUMBER_UNDEFINED_TYPES);
-
-      // clear the set so only new undefined types will be logged
-      undefinedTypes.clear();
+    if (problemsToFilePaths.isEmpty()) {
+      return;
+    }
+    javaClasspath.logSuspiciousEmptyLibraries();
+    if (!isAutoScan()) {
+      // In autoscan, test + main code are analyzed in the same batch, and we do not make the distinction between
+      // test and main libraries, everything is inside "sonar.java.libraries", it is expected to let the test property empty.
+      javaTestClasspath.logSuspiciousEmptyLibraries();
     }
+    logUndefinedTypes(LOGGED_MAX_NUMBER_UNDEFINED_TYPES);
+
+    // clear the set so only new undefined types will be logged
+    problemsToFilePaths.clear();
   }
 
   private void logUndefinedTypes(int maxLines) {
     logParserMessages(
-      undefinedTypes.stream()
-        .filter(m -> m.type() == JProblem.Type.UNDEFINED_TYPE),
+      problemsToFilePaths.entrySet().stream()
+        .filter(entry -> entry.getKey().type() == JProblem.Type.UNDEFINED_TYPE),
       maxLines,
       "Unresolved imports/types have been detected during analysis. Enable DEBUG mode to see them.",
       "Unresolved imports/types:"
     );
     logParserMessages(
-      undefinedTypes.stream()
-        .filter(m -> m.type() == JProblem.Type.PREVIEW_FEATURE_USED),
+      problemsToFilePaths.entrySet().stream()
+        .filter(entry -> entry.getKey().type() == JProblem.Type.PREVIEW_FEATURE_USED),
       maxLines,
       "Use of preview features have been detected during analysis. Enable DEBUG mode to see them.",
       "Use of preview features:"
     );
   }
 
-  private static void logParserMessages(Stream<JProblem> messages, int maxLines, String warningMessage, String debugMessage) {
-    final List<String> messagesList = messages
-      .map(Object::toString)
-      .sorted()
+  private static void logParserMessages(Stream<Map.Entry<JProblem, List<String>>> messages, int maxProblems, String warningMessage, String debugMessage) {
+    final String problemDelimiter = System.lineSeparator() + "- ";
+    final List<List<String>> messagesList = messages
+      .sorted(Comparator.comparing(entry -> entry.getKey().toString()))
+      .limit(maxProblems + 1L)
```

Comment:
Why `+1` here? On line 561 we don't do it.

Reply:
The `+1L` here allows us to prepare messages for up to `maxProblems` JProblems.
Because we do not know how many elements are contained in the stream but we know that we want at most `maxProblems`, this trick allows us to apply the mapping function unnecessarily to at most 1 element.
Knowing that we have one extra element allows us to decide whether we want to mention in the debug message that we are truncating the logs.

Reply:
The extra comment is OK for me! Thanks for the rationale.